robocin/utility/README.md

```diff
@@ -0,0 +1,111 @@
+# utility
+
+A collection of utility functions and classes.
+
+## Table of Contents
+
+- [concepts](#concepts)
+- [epsilon](#epsilon)
+- [fuzzy_compare](#fuzzy_compare)
+- [type_traits](#type_traits)
+
+<a name="concepts"></a>
+
+## [`concepts`](concepts.h)
+
+The [concepts](concepts.h) header provides a set of concepts extended the standard library.
+
+- `arithmetic`: a concept that represents a type that is either an integral or floating point type;
+
+<a name="epsilon"></a>
+
+## [`epsilon`](epsilon.h.in)
+
+The [epsilon](epsilon.h.in) is a [CMake configure file](https://cmake.org/cmake/help/latest/command/configure_file.html)
+that defines the epsilon value passed to CMake during project build through the `ROBOCIN_FLOAT_EPSILON`,
+`ROBOCIN_DOUBLE_EPSILON` and `ROBOCIN_LONG_DOUBLE_EPSILON` flags (see [CMakeLists.txt](CMakeLists.txt)).
+
+These values will be used by the library to perform real number comparisons.
+
+* The existence of these values can be verified as follows:
+
+```cpp
+robocin::has_epsilon<float>::value;
+robocin::has_epsilon<double>::value;
+robocin::has_epsilon<long double>::value;
+```
+
+or:
+
+```cpp
+robocin::has_epsilon_v<float>;
+robocin::has_epsilon_v<double>;
+robocin::has_epsilon_v<long double>;
+```
+
+* When provided, you can access them without using:
+
+```cpp
+robocin::has_epsilon<float>::epsilon;
+robocin::has_epsilon<double>::epsilon;
+robocin::has_epsilon<long double>::epsilon;
+```
+
+or:
+
+```cpp
+robocin::epsilon_v<float>;
+robocin::epsilon_v<double>;
+robocin::epsilon_v<long double>;
+```
+
+<a name="fuzzy_compare"></a>
+
+## [`fuzzy_compare`](fuzzy_compare.h)
+
+The [fuzzy_compare](fuzzy_compare.h) header provides a set of functions to perform fuzzy comparisons and functor classes
+to be used with the standard library algorithms. There are many ways to compare real numbers, but for the applications
+that RobôCIn and other robotics teams use, the way chosen is to compare the absolute difference against
+a [tolerance margin](#epsilon):
+
+```cpp
+// check if a floating point number 'is zero':
+std::abs(value) <= epsilon;
+
+// check if two floating point numbers 'are equal':
+std::abs(lhs - rhs) <= epsilon;
+```
```

Comment:
@bcs5, alguns links que cheguei a ler antes de desenvolver:
- [Qt Compare Two Floats](https://embeddeduse.com/2019/08/26/qt-compare-two-floats);
- [The floating-point guide](https://floating-point-gui.de/errors/comparison);
- [How should I do floating point comparison?](https://stackoverflow.com/a/32334103);
- [Comparing Floating Point Numbers, 2012 Edition](https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition);

Após a leitura pensei em fazer algum método que levasse em consideração mais casos, mas a maioria implica em mais complexidade em uma função que deveria ser simples, e talvez, pra nossa aplicação uma comparação extremamente precisa não se faça necessário...
Busquei aquilo que equipes mais tradicionais fazem, e consiste no mesmo comparativo de diferença absoluta para uma margem:
* TIGERs: [`isEqual`](https://github.com/TIGERs-Mannheim/Sumatra/blob/cb8ea325f3e5a1dcc719415fe697108e4705c8b7/modules/common-math/src/main/java/edu/tigers/sumatra/math/SumatraMath.java#L140) e [`isZero`](https://github.com/TIGERs-Mannheim/Sumatra/blob/cb8ea325f3e5a1dcc719415fe697108e4705c8b7/modules/common-math/src/main/java/edu/tigers/sumatra/math/SumatraMath.java#L168);

* ER-Force: [`std::abs(xNext - yNext) < SAME_POINT_EPSILON`](https://github.com/robotics-erlangen/framework/blob/0f5ffffa4417be583357c923ee4217183ed5c260/src/amun/strategy/path/speedprofile.cpp#L243);

Reply:
Alguma motivação do uso de fuzzy no prefixo? Só havia ouvido isso no contexto de strings. Para floating points melhor normalizar(ser o padrão) o uso, e só esquecer que há comparação direta.

Reply:
A motivação inicial é que framework que utilizávamos antigamente tinha métodos com esse nome documentados ([link](https://doc.qt.io/qt-5/qtglobal.html#qFuzzyCompare)). Posteriormente fazendo uma [busca mais aprofundada](https://scholar.google.com/scholar?hl=pt-BR&as_sdt=0%2C5&q=fuzzy+compare+real+numbers), vi que é um nome que se aplica pra esse uso, mas também a princípio o que mais aparece é relativo à strings.